developer_guides/topology2/topology2.rst

```diff
@@ -0,0 +1,895 @@
+			!deprecated [
+				"preload_count"
+			]
+
+			#
+			# pga widget objects instantiated within the same alsaconf node must have
+			# unique instance attribute
+			#
+			unique	"instance"
+		}
+
+		# default attribute values
+		type 		"pga"
+		uuid 		"7e:67:7e:b7:f4:5f:88:41:af:14:fb:a8:bd:bf:86:82"
+	}
+
+1.1.5 Attribute Constraints
+***************************
+One of the key features of Topology 2.0 is validation of the values provided for objects. This is achieved
+with the help of constraints added to the attribute definition. Constraints can be added to an attribute using
+the "constraints" keyword as follows:
+
+.. code-block:: bash
+
+	DefineAttribute."foo" {
+		constraints {}
+	}
+
+Currently, 3 types of constriaints are supported:
+
+* min: min value for attribute, applicable to integer type attributes
+* max: max value for attribute, applicable to integer type attributes
+
+  For example, the pga class definition can be expanded with an attribute for "ramp_step_ms" with min and
+  max values as follows:
+
+  .. code-block:: bash
+
+	DefineAttribute."ramp_step_ms" {
+		constraints {
+			min 200
+			max 500
+		}
+	}
+
+* valid values: an array of acceptable human readable values, applicable to string type attributes.
+
+  For example, the pga class can have an attribue for "ramp_step_type" with pre-defined values as follows:
+  
+  .. code-block:: bash
+  
+  	DefineAttribute."ramp_step_type" {
+		type	"string"
+		constraints {
+			!valid_values [
+				"linear"
+				"log"
+				"linear_zc"
+				"log_zc"
+			]
+		}
+	}
+
+1.1.6 Attributes with token references
+**************************************
+Classes can also contain attribute definitions with the token_ref node which contains the name of the tuple array
+that the attribute should be built into. The tuple arrays are added to the private data section of the object that
+is being built. For example, both the ramp_step_ms and ramp_step_type attributes in the pga class contain the
+token_ref "sof_tkn_volume.word" indicating that these attributes should be packed with the sof_tkn_volume tuple array
+of type "word".
+
+.. code-block:: bash
+
+		#
+		# Volume ramp step in milliseconds
+		#
+		DefineAttribute."ramp_step_ms" {
+			# Token set reference name
+			token_ref	"sof_tkn_volume.word"
+			constraints {
+				min 200
+				max 500
+			}
+		}
+		DefineAttribute."ramp_step_type" {
+			type	"string"
+			# Token set reference name
+			token_ref	"sof_tkn_volume.word"
+			constraints {
+				!valid_values [
+					"linear"
+					"log"
+					"linear_zc"
+					"log_zc"
+				]
+			}
+		}
+
+Sometimes, valid_values for attributes might need to be translated from the human readable values to integer tuple values so
+that it can be parsed correctly by the kernel driver. In the example above, valid values for ramp_step_type are defined
+as hunam reabled string values such as linera, log etc. which are translated to tuple values 0, 1, etc respectively before 
+getting added to the tuple array.
```

Comment:
not completely following, so the kernel would parse 0..3 and pass the value to firmware without having to deal with an intermediate dictionary with string->integer conversions?